feat: add session handling for Kotlin and Swift (#24)
amplitude/amplitude-dev-center: includes/sdk-next-gen-mobile-session.md

When the app enters the foreground, Amplitude tracks a session start, and starts a countdown based on `setMinTimeBetweenSessionsMillis()`. Amplitude extends the session and restarts the countdown any time it tracks a new event. If the countdown expires, Amplitude waits until the next event to track a session end event.

Amplitude doesn't set user properties on session events by default. To add these properties, use `identify()` and `setUserId()`. Amplitude aggregates the user property state and associates the user with events based on `device_id` or `user_id`.

Due to the way in which Amplitude manages sessions, there are scenarios where the SDK works expected but it may appear as if events are missing or session tracking is inaccurate:

* If a user doesn't return to the app, Amplitude does not track a session end event to correspond with a session start event.
* If you track an event in the background, it's possible that Amplitude perceives the session length to be longer than the user spends on the app in the foreground.

> * If you modify user properties between the last event and the session end event, the session end event reflects the updated user properties, which may differ from other properties associated with events in the same session. To address this, use an enrichment plugin to set `event['$skip_user_properties_sync']` to `true` on the session end event, which prevents Amplitude from synchronizing properties for that specific event. To learn more about [$skip_user_properties_sync](/data/sdks/flutter/#skip_user_properties_sync)

Reviewer: ```suggestion
* If you modify user properties between the last event and the session end event, the session end event reflects the updated user properties, which may differ from other properties associated with events in the same session. To address this, use an enrichment plugin to set `event['$skip_user_properties_sync']` to `true` on the session end event, which prevents Amplitude from synchronizing properties for that specific event. See [$skip_user_properties_sync](/data/sdks/flutter/#skip_user_properties_sync) in the Flutter SDK article to learn more.
```

> Amplitude starts a session when the app is brought into the foreground or when an event is tracked in background.

Reviewer: ```suggestion
Amplitude starts a session when the app is brought into the foreground or when an event is tracked in the background.
```

> Amplitude tracks a session start event upon entering the foreground and starts a countdown set by `setMinTimeBetweenSessionsMillis()`. If a new event is tracked, Amplitude extends the session and restart the countdown with the same duration. If the countdown expires, Amplitude doesn't immediately track a session end event but  waits until the next event.

Reviewer: ```suggestion
When the app enters the foreground, Amplitude tracks a session start, and starts a countdown based on `setMinTimeBetweenSessionsMillis()`. Amplitude extends the session and restarts the countdown any time it tracks a new event. If the countdown expires, Amplitude waits until the next event to track a session end event.
```

> Amplitude tracks a session start event upon entering the foreground and starts a countdown set by `setMinTimeBetweenSessionsMillis()`. If a new event is tracked, Amplitude extends the session and restart the countdown with the same duration. If the countdown expires, Amplitude doesn't immediately track a session end event but  waits until the next event.

Reviewer 2: Can you please clarify that the `setMinTimeBetweenSessionsMillis(). e.g. Sessions will continue for the entire time the app is in the foreground, if the app is in the background for more than "min time between sessions" a new session will be started the next time the user opens the app, or tracks an event in the background.

Author: Change at https://github.com/amplitude/amplitude-dev-center/pull/24/commits/95a11e43b07a65ed091268f3948358c1270f61df. WDYT?

> Amplitude SDKs don't set user properties to session events directly. Instead `identify()` and `setUserId()` update user properties from the SDK. The user property state is then aggregated on the backend and associated with events during ingestion based on `device_id` or `user_id`.

Reviewer: ```suggestion
Amplitude doesn't set user properties on session events by default. To add these properties, use `identify()` and `setUserId()`. Amplitude aggregates the user property state and associates the user with events based on `device_id` or `user_id`.
```

> If they would like to not sync user properties on these events we have an option of setting "$skip_user_properties_sync": true on the event, and this will stop us from syncing properties on that event in particular.

Reviewer: ```suggestion
If you don't want to sync user properties with session events, set the `$skip_user_properties_sync` property to `true` to instruct Amplitude to stop syncing properties on that particular event.
```
@Mercy811 Where would someone set $skip_user_properties_sync?

Author: Make it up at https://github.com/amplitude/amplitude-dev-center/pull/24/commits/fea9c627ae0b117cff48d169fee5f81f79aa9ba9

> Due to how Amplitude manages sessions, there are certain scenarios where the SDK works expected but may appear as if events are missing or session tracking is inaccurate:

Reviewer: ```suggestion
Due to the way in which Amplitude manages sessions, there are scenarios where the SDK works expected but it may appear as if events are missing or session tracking is inaccurate:
```

> * If a user never returns to the app , there will be no corresponding session end event for the session start event

Reviewer: ```suggestion
* If a user doesn't return to the app, Amplitude does not track a session end event to correspond with a session start event.
```

> * If you disable session tracking and only track few events in the app, the tracked session length will be shorter than the time the user actually spends on the app

Reviewer: ```suggestion
* If you disable session tracking and track a minimal number of events, Amplitude perceives the session length to be shorter than the user spends on the app.
```

> * If you modify user properties between the last event and the session end event, the session end event will reflect the updated user properties, potentially differing from other properties associated with events in the same session. To address this, you can utilize an enrichment plugin to set `$skip_user_properties_sync` to true on the session end event, preventing the backend from synchronizing properties for that specific event

Reviewer: ```suggestion
* If you modify user properties between the last event and the session end event, the session end event reflects the updated user properties, which may differ from other properties associated with events in the same session. To address this, use an enrichment plugin to set `$skip_user_properties_sync` to `true` on the session end event, which prevents Amplitude from synchronizing properties for that specific event.

```
